src/green/builder.rs

```diff
@@ -26,22 +70,40 @@ impl NodeCache {
         // For `libsyntax/parse/parser.rs`, measurements show that deduping saves
         // 17% of the memory for green nodes!
         // Future work: make hashing faster by avoiding rehashing of subtrees.
-        if node.children().len() <= 3 {
-            match self.nodes.get(&node) {
-                Some(existing) => node = existing.clone(),
-                None => assert!(self.nodes.insert(node.clone())),
+        if num_children <= MAX_CHILDREN {
+            let mut hash = GreenNodeHash::new(kind);
+            let mut collected_children =
+                SmallVec::<[GreenElement; MAX_CHILDREN]>::with_capacity(MAX_CHILDREN);
+            for child in children {
+                collected_children.push(child);
+            }
```

Comment:
This should be writable with a `collect` (i.e. `let collected_children: SmallVec::<[_; MAX_CHILDREN]> = children.collect()`).